File: RootTools2/src/main/java/com/rarnu/tools/neo/xposed/ads/FuckMusic.java
package com.rarnu.tools.neo.xposed.ads;

import android.app.Activity;
import android.content.Context;
import android.net.Uri;
import android.view.View;
import android.view.ViewGroup;
import com.rarnu.tools.neo.xposed.XpUtils;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.callbacks.XC_LoadPackage;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.List;

/**
 * Created by rarnu on 11/18/16.
 */
public class FuckMusic {

    public static void fuckMusic(final XC_LoadPackage.LoadPackageParam loadPackageParam) {
        Class<?> clsListener = XpUtils.findClass(loadPackageParam.classLoader, "com.android.volley.Response$Listener");
        Class<?> clsErrorListener = XpUtils.findClass(loadPackageParam.classLoader, "com.android.volley.Response$ErrorListener");
        final Class<?> clsAdInfo = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.player.util.AdUtils$AdInfo");

        XpUtils.findAndHookMethod("com.miui.player.util.AdUtils", loadPackageParam.classLoader, "isAdEnable", XC_MethodReplacement.returnConstant(false));
        if (clsListener != null && clsErrorListener != null) {
            XpUtils.findAndHookMethod("com.miui.player.util.AdUtils", loadPackageParam.classLoader, "getPlayAd", clsListener, clsErrorListener, XC_MethodReplacement.returnConstant(null));
        }
        XpUtils.findAndHookMethod("com.miui.player.util.ExperimentsHelper", loadPackageParam.classLoader, "isAdEnabled", XC_MethodReplacement.returnConstant(false));
        if (clsAdInfo != null) {
            XpUtils.findAndHookMethod("com.miui.player.util.AdUtils", loadPackageParam.classLoader, "handleDeepLinkUrl", Activity.class, clsAdInfo, XC_MethodReplacement.returnConstant(null));
            XpUtils.findAndHookMethod("com.miui.player.util.AdUtils", loadPackageParam.classLoader, "showAlertAndDownload", Activity.class, clsAdInfo, XC_MethodReplacement.returnConstant(null));
            XpUtils.findAndHookMethod("com.miui.player.util.AdUtils", loadPackageParam.classLoader, "handleAdClick", Activity.class, clsAdInfo, XC_MethodReplacement.returnConstant(null));
            XpUtils.findAndHookMethod("com.miui.player.util.AdUtils", loadPackageParam.classLoader, "postPlayAdStat", String.class, clsAdInfo, XC_MethodReplacement.returnConstant(null));
        }
        XpUtils.findAndHookMethod("com.miui.player.phone.view.NowplayingAlbumPage", loadPackageParam.classLoader, "getPlayAd", XC_MethodReplacement.returnConstant(null));
        XpUtils.findAndHookMethod("com.miui.player.util.Configuration", loadPackageParam.classLoader, "isCmTest", XC_MethodReplacement.returnConstant(true));
        XpUtils.findAndHookMethod("com.miui.player.util.Configuration", loadPackageParam.classLoader, "isCpLogoVisiable", XC_MethodReplacement.returnConstant(false));

        // fuck the ad under account
        Class<?> clsAdInfo2 = XpUtils.findClass(loadPackageParam.classLoader, "com.xiaomi.music.online.model.AdInfo");
        if (clsAdInfo != null) {
            XpUtils.findAndHookMethod("com.miui.player.util.AdForbidManager", loadPackageParam.classLoader, "recordAdInfo", clsAdInfo2, int.class, XC_MethodReplacement.returnConstant(null));
            XpUtils.findAndHookMethod("com.miui.player.util.AdForbidManager", loadPackageParam.classLoader, "addForbidInfo", String.class, XC_MethodReplacement.returnConstant(null));
            XpUtils.findAndHookMethod("com.miui.player.util.AdForbidManager", loadPackageParam.classLoader, "isForbidden", String.class, XC_MethodReplacement.returnConstant(true));
        }
        XpUtils.findAndHookMethod("com.miui.player.hybrid.feature.GetAdInfo", loadPackageParam.classLoader, "addAdQueryParams", Context.class, Uri.class, XC_MethodReplacement.returnConstant(""));

        XpUtils.findAndHookMethod("com.miui.player.display.view.cell.BannerAdItemCell", loadPackageParam.classLoader, "onFinishInflate", new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                Class<?> clsThis = param.thisObject.getClass();
                View vThis = (View) param.thisObject;
                Field fText = clsThis.getDeclaredField("mClose");
                fText.setAccessible(true);
                View vText = (View) fText.get(param.thisObject);
                vText.setVisibility(View.GONE);
                Field fImg = clsThis.getDeclaredField("mImage");
                fImg.setAccessible(true);
                View vImg  = (View) fImg.get(param.thisObject);
                vImg.setVisibility(View.GONE);
                ViewGroup.LayoutParams lp = vThis.getLayoutParams();
                lp.height = 0;
                vThis.setLayoutParams(lp);
            }
        });

        // try to remove copy right validation, removed.

        /*

        final Class<?> clsOnQualitySelectedListener = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.player.vip.DownloadVipHelper$OnQualitySelectedListener");
        final Class<?> clsResult = XpUtils.findClass(loadPackageParam.classLoader, "com.xiaomi.music.model.Result");
        if (clsOnQualitySelectedListener != null && clsResult != null) {
            XpUtils.findAndHookMethod("com.miui.player.vip.DownloadVipHelper", loadPackageParam.classLoader, "handleBatchDownloadPermission", Activity.class, List.class, int.class, boolean.class, clsOnQualitySelectedListener, clsResult, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    Object oResult = param.args[5];
                    Field fErrorCode = clsResult.getDeclaredField("mErrorCode");
                    fErrorCode.setAccessible(true);
                    fErrorCode.setInt(oResult, 1);
                }
            });
        }

        final Class<?> clsDownloadOne = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.player.download.MusicDownloader$DownloadOne");
        if (clsOnQualitySelectedListener != null && clsResult != null && clsDownloadOne != null) {
            XpUtils.findAndHookMethod("com.miui.player.vip.DownloadVipHelper", loadPackageParam.classLoader, "handleMusicUrl", Activity.class, clsDownloadOne, int.class, boolean.class, clsOnQualitySelectedListener, clsResult, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    Object oResult = param.args[5];
                    Field fErrorCode = clsResult.getDeclaredField("mErrorCode");
                    fErrorCode.setAccessible(true);
                    fErrorCode.setInt(oResult, 1);
                }
            });
        }

        // TODO: to be removed....
        if (clsDownloadOne != null) {
            XpUtils.findAndHookMethod("com.miui.player.vip.DownloadVipHelper", loadPackageParam.classLoader, "getMusicUrl", Activity.class, clsDownloadOne, int.class, new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) throws Throwable {
                    Object oDownloadOne = param.args[1];
                    Field fValue = clsDownloadOne.getDeclaredField("mValue");
                    fValue.setAccessible(true);
                    Object oValue = fValue.get(oDownloadOne);
                    Field fGlobalId = oValue.getClass().getDeclaredField("mGlobalId");
                    fGlobalId.setAccessible(true);
                    String globalId = (String) fGlobalId.get(oValue);
                    try {
                        Class<?> clsEngineHelper = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.player.util.EngineHelper");
                        Method mGet = clsEngineHelper.getDeclaredMethod("get", Context.class);
                        Object oMusicEngine = mGet.invoke(null, param.args[0]);
                        Method mGetOnlineEngine = oMusicEngine.getClass().getDeclaredMethod("getOnlineEngine");
                        Object oOnlineEngine = mGetOnlineEngine.invoke(oMusicEngine);
                        Method mGetMusicLink = oOnlineEngine.getClass().getDeclaredMethod("getMusicLink", Context.class, String.class, int.class, int.class);

                        Class<?> clsGlobalId = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.player.content.GlobalIds");
                        Method mGetId = clsGlobalId.getDeclaredMethod("getId", String.class);
                        String newGlobalId = (String) mGetId.invoke(null, globalId);

                        Class<?> clsQualityUtils = XpUtils.findClass(loadPackageParam.classLoader, "com.miui.player.util.QualityUtils");
                        Method mToBitrate = clsQualityUtils.getDeclaredMethod("toBitrate", int.class);
                        int bitrate = (int) mToBitrate.invoke(null, param.args[2]);

                        Object oResult = mGetMusicLink.invoke(oOnlineEngine, param.args[0], newGlobalId, bitrate, 1);
                        XposedBridge.log(String.format("getMusicUrl => %s", oResult.toString()));

                    } catch (Exception e) {

                    }
                }
            });
        }
        XpUtils.findAndHookMethod("com.xiaomi.music.online.model.BatchDownloadPermission", loadPackageParam.classLoader, "allDownloadAllowed", XC_MethodReplacement.returnConstant(true));

        */
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/xposed/ads/FuckMusic.kt
@file:Suppress("Duplicates")

package com.rarnu.tools.neo.xposed.ads

import android.app.Activity
import android.content.Context
import android.net.Uri
import android.util.AttributeSet
import android.view.View
import com.rarnu.xfunc.*
import de.robv.android.xposed.XposedHelpers

/**
 * Created by rarnu on 11/18/16.
 */
object FuckMusic {

    fun fuckMusic(pkg: XposedPkg) {
        val clsListener = pkg.findClass("com.android.volley.Response\$Listener")
        val clsErrorListener = pkg.findClass("com.android.volley.Response\$ErrorListener")
        val clsAdInfo = pkg.findClass("com.miui.player.util.AdUtils\$AdInfo")
        pkg.findClass("com.miui.player.util.AdUtils").apply {
            findMethod("isAdEnable").hook { replace { result = false } }
            if (clsListener != null && clsErrorListener != null) {
                findMethod("getPlayAd", clsListener, clsErrorListener).hook { replace { result = null } }
            }
            if (clsAdInfo != null) {
                findMethod("handleDeepLinkUrl", Activity::class.java, clsAdInfo).hook { replace { result = null } }
                findMethod("showAlertAndDownload", Activity::class.java, clsAdInfo).hook { replace { result = null } }
                findMethod("handleAdClick", Activity::class.java, clsAdInfo).hook { replace { result = null } }
                findMethod("postPlayAdStat", String::class.java, clsAdInfo).hook { replace { result = null } }
            }
        }

        pkg.findClass("com.miui.player.util.ExperimentsHelper").findMethod("isAdEnabled").hook { replace { result = null } }
        pkg.findClass("com.miui.player.phone.view.NowplayingAlbumPage").findMethod("getPlayAd").hook { replace { result = null } }
        pkg.findClass("com.miui.player.util.Configuration").apply {
            findMethod("isCmTest").hook { replace { result = true } }
            findMethod("isCpLogoVisiable").hook { replace { result = false } }
        }

        // fuck the ad under account
        val clsAdInfo2 = pkg.findClass("com.xiaomi.music.online.model.AdInfo")
        if (clsAdInfo != null) {
            pkg.findClass("com.miui.player.util.AdForbidManager").apply {
                findMethod("recordAdInfo", clsAdInfo2, Integer.TYPE).hook { replace { result = null } }
                findMethod("addForbidInfo", String::class.java).hook { replace { result = null } }
                findMethod("isForbidden", String::class.java).hook { replace { result = true } }
            }
        }
        pkg.findClass("com.miui.player.hybrid.feature.GetAdInfo").findMethod("addAdQueryParams", Context::class.java, Uri::class.java).hook { replace { result = "" } }
        pkg.findClass("com.miui.player.display.view.cell.BannerAdItemCell").findMethod("onFinishInflate").hook {
            after {
                val vThis = thisObject as View
                try {
                    (XposedHelpers.getObjectField(thisObject, "mClose") as View?)?.visibility = View.GONE
                } catch (t: Throwable) {
                }
                try {
                    (XposedHelpers.getObjectField(thisObject, "mImage") as View?)?.visibility = View.GONE
                } catch (t: Throwable) {
                }
                vThis.layoutParams = vThis.layoutParams.apply { height = 0 }
            }
        }

        // 2.7.300
        pkg.findClass("com.miui.player.content.MusicHybridProvider").findMethod("parseCommand", String::class.java).hook {
            before {
                val scheme = args[0] as String?
                if (scheme == "advertise") {
                    args[0] = ""
                }
            }
        }
        pkg.findClass("com.miui.systemAdSolution.landingPage.LandingPageService").findMethod("init", Context::class.java).hook { replace { result = null } }

        // 2.7.400
        pkg.findClass("com.miui.player.phone.view.NowplayingContentView").findMethod("setInfoVisibility", java.lang.Boolean.TYPE).hook {
            before {
                args[0] = true
            }
        }

        if (clsAdInfo2 != null) {
            pkg.findClass("com.miui.player.phone.view.NowplayingContentView\$ShowAdRunnable").findConstructor(clsAdInfo2, java.lang.Boolean.TYPE).hook {
                before {
                    args[1] = false
                }
            }
        }
        pkg.findClass("com.miui.player.phone.view.NowplayingContentView\$ShowAdRunnable").findMethod("setLoadAd", java.lang.Boolean.TYPE).hook {
            before {
                args[0] = false
            }
        }

        // 2.8
        val clsDisplayItem = pkg.findClass("com.miui.player.display.model.DisplayItem")
        if (clsDisplayItem != null) {
            pkg.findClass("com.miui.player.display.view.SearchPopularKeyCard").findMethod("onBindItem", clsDisplayItem, Integer.TYPE).hook {
                after {
                    (thisObject as View).apply {
                        layoutParams.height = 0
                        visibility = View.GONE
                    }
                }
            }
            pkg.findClass("com.miui.player.display.view.BannerCard").findMethod("onBindItem", clsDisplayItem, Integer.TYPE).hook {
                after {
                    (thisObject as View).apply {
                        layoutParams.height = 0
                        visibility = View.GONE
                    }
                }
            }
        }

        val clsAdShowingListener = pkg.findClass("com.miui.player.phone.view.NowplayingAlbumPage\$AdShowingListener")
        if (clsAdShowingListener != null) {
            pkg.findClass("com.miui.player.phone.view.NowplayingAlbumView").findMethod("setAdShowingListener", clsAdShowingListener).hook { replace { result = null } }
        }
        pkg.findClass("com.miui.player.phone.view.NowplayingAlbumPage\$ShowAdRunnable").findConstructor(java.lang.Boolean.TYPE).hook {
            before {
                args[0] = false
            }
        }

        // 2.9.0
        pkg.findClass("com.miui.player.content.preference.PreferenceCache").findMethod("getBoolean", Context::class.java, String::class.java).hook {
            after {
                if (args[1] == "ad_recommend") {
                    result = false
                }
            }
        }
        pkg.findClass("com.miui.player.util.ExperimentsHelper").findMethod("isAdEnabled").hook { replace { result = false } }
        pkg.findClass("com.miui.player.phone.view.NowplayingAlbumPage").findConstructor(Context::class.java, AttributeSet::class.java, Integer.TYPE).hook {
            after {
                val mAdMark = XposedHelpers.getObjectField(thisObject, "mAdMark") as View?
                mAdMark?.visibility = View.GONE
            }
        }
    }

}
